packages/server/src/transformer.ts

```diff
@@ -86,9 +89,113 @@ export type DefaultDataTransformer = CombinedDataTransformer & {
 
 /**
  * @internal
+ * todo: This is not used if TSON is enabled - does the default TSON transformer need a `_default` flag?
  */
 export const defaultTransformer: DefaultDataTransformer = {
   _default: true,
   input: { serialize: (obj) => obj, deserialize: (obj) => obj },
   output: { serialize: (obj) => obj, deserialize: (obj) => obj },
 };
+
+export type TrpcTuplesonConfig =
+  | TsonAsyncOptions
+  | {
+      input: TsonAsyncOptions;
+      output: TsonAsyncOptions;
+    };
+
+/**
+ * @internal
+ */
+export function getDataTransformer<
+  TOpts extends {
+    transformer?: DataTransformerOptions | undefined;
+    experimental_tuplesonOptions?: TrpcTuplesonConfig;
+  },
+>(opts: TOpts | undefined) {
+  const transformer = opts?.experimental_tuplesonOptions
+    ? getTsonTransformer(opts)
+    : getLegacyTransformer(opts);
+
+  return transformer;
+}
+
+function getLegacyTransformer<TTransformer extends DataTransformerOptions>(
+  opts: { transformer?: TTransformer } | undefined,
+): CombinedDataTransformer {
+  const transformer = opts?.transformer;
+  if (!transformer) {
+    return defaultTransformer;
+  }
+  if ('input' in transformer) {
+    return transformer;
+  }
+  return {
+    input: transformer,
+    output: transformer,
+  };
+}
+
+function getTsonTransformer<
+  TOpts extends {
+    experimental_tuplesonOptions?: TrpcTuplesonConfig | undefined;
+    transformer?: DataTransformerOptions | undefined;
+  },
+>(opts: TOpts | undefined) {
+  if (!opts?.experimental_tuplesonOptions) {
+    throw new Error(
+      'TSON is not enabled. This should not happen; it is a bug.',
+    );
+  }
+
+  const officialTsonSyncHandlers = [
+    tsonBigint,
+    tsonDate,
+    tsonRegExp,
+    tsonSet,
+    tsonMap,
+    tsonUndefined,
+    tsonNumberGuard,
+    tsonURL,
+    tsonSymbol,
+  ] as const;
```

Comment:
I am not sure we should be adding a bunch of transformers automatically as it is quite a big change,

I'd rather have only support for Promise and AsyncIterable and let the user define the others

Reply:
So, if I understand, were basically just defaulting to JSON for sync types, since tupleson only tuplefies non-json-serializables? By the way, I missed that little detail for way too long, and may have made some assumptions that rely on ALL serialized values being wrapped in a tuple (with nonce)